(Design) Formatted Parts (#463)
unicode-org/message-format-wg: exploration/0003-formatted-parts.md

# Formatted Parts

Status: **Proposed**

<details>
	<summary>Metadata</summary>
	<dl>
		<dt>Contributors</dt>
		<dd>@eemeli</dd>
		<dt>First proposed</dt>
		<dd>2023-08-29</dd>
		<dt>Pull Request</dt>
		<dd>#000</dd>
	</dl>
</details>

## Objective

Messages often include structures and values
that are not necessarily best represented by a plain concatenated string.

It would be useful to define a formatted-parts target for MessageFormat 2.

## Background

Past examples have shown us that if we don't provide a formatter to parts,
the string output will be re-parsed and re-processed by users.

> ## Use-Cases

Author: @aphillips This section has been expanded since this comment. Sufficiently?

  The JS implementation of the MF2 formatted-parts representation should be able to match their structure,
  at least as far as that's possible and appropriate.

## Proposed Design

The formatted-parts API is included in the spec as an optional but recommended formatting target.

The shape of the formatted-parts output is defined in a manner similar to the data model,
which includes TypeScript, JSON Schema, and XML DTD definitions of the same data structure.

At the top level, the formatted-parts result is an iterable sequence of parts.
Parts corresponding to each _text_ can be simpler than those of _expressions_,
as they do not have a `source` other than their `value`,
or set any of the other possible metadata fields.

```ts
type MessageParts = Iterable<
  MessageTextPart | MessageExpressionPart | MessageBiDiIsolationPart
>;

interface MessageTextPart {
  type: "text";
  value: string;
}
```

For MessageExpressionPart, the `source` corresponds to the expression's fallback value.
The parts' `dir` and `locale` may be defined if overridden by an expression attribute or set otherwise.

> Each part should have at most one of `value` or `parts` defined;
> some may have none.

Reviewer: Does this make sense? Or would it be better to say:

```suggestion
Each part MUST have either a `value` or `parts` defined. 
A part MAY have a `value` that is the empty string.
A part MAY have a `parts` that is an empty list.
```

Author: The suggestion would make the proposed MessageFallbackPart invalid, as it does not include either `value` or `parts`. It's conceivable for other parts to also exist which do not include either, such as open/close expressions without an annotation.

Reviewer: Would an empty `value` or empty `parts` satisfy that? Or a fallback could have a string expression? Empty strings don't result in the erroneous emission of the string `null` 😉 

I understand that it would "break" the current definition: we should decide what the shapes should be and make consistent.

Author: For fallback when formatting to a string the `{...}` make sense as a visual indicator, but for a formatted-parts consumer some different representation could be better. So including an explicit `value` or `parts` would be misleading.

For open/close, it doesn't make sense to define their explicit parts shapes in this spec, but for JS I have them as:

```ts
interface MessageMarkupPart {
  type: 'open' | 'close';
  source: string;
  name: string;
  value?: unknown;
  options: { [key: string]: unknown };
}
```

There, the `value` would be `'b'` for `{b +html}`, but it would not be set for `{+html.b}`. Setting it to an empty string would be misleading, as `{+foo}` and `{|| +foo}` could easily mean different things.

Reviewer 2: In https://github.com/unicode-org/message-format-wg/pull/463#discussion_r1406427143 I'm suggesting to define separate interfaces for single-valued and multi-valued parts. This could extend to fallback parts and markup, as well.

```ts
interface MessageExpressionPart {
  type: string;
  source: string;
  parts?: Iterable<{ type: string; value: unknown }>;
  value?: unknown;
  dir?: "ltr" | "rtl" | "auto";
  locale?: string;
}
```

The bidi isolation strategies included in the spec may require
the insertion of MessageBiDiIsolationParts in the formatted-parts output.

```ts
interface MessageBiDiIsolationPart {
  type: "bidiIsolation";
  value: "\u2066" | "\u2067" | "\u2068" | "\u2069"; // LRI | RLI | FSI | PDI
}

> ```

Reviewer: Maybe? But since the parts have a direction, the bidi isolation parts are not necessarily required. Let's consider pattern:

```
{Today is {$date :datetime}}
```

In locale `ar-AE`, the output of the `:datetime` function might be "30‏/08‏/2023" with a `dir` of `rtl`. Note that the string looks bad here because it is an LTR context and not bidi isolated. There are U+200F RLMs after `08` and `30` to assist, but the string needs to be wrapped.

Ignoring that the date has interior parts, getting a MessageExpressionPart:

```
value: "30‏/08‏/2023"
dir: "rtl"
```

... could result in a formatted string (in an HTML context) like:

```
Today is <span dir=rtl>30‏/08‏/2023</span>.
```

This draws correctly without any further intervention.

If this became a list of parts it still works:

```
parts: [
   { name: "day", value: "30\u200f" },
   { name: "sep", value: "/"},
   { name: "month", value: "08\u200f"},
   { name: "sep", value: "/"},
   { name: "year", value: "2023"}
]
dir: "rtl"
```

The caller could use this to produce:

```html
Today is <span dir=rtl>
<span class="day">30&#x200f;</span>/08&#x200f;2023</span>.
```

[Test](https://w3c.github.io/i18n-discuss/explainers/bidi-html-demo.html?text=Today%20is%20%3Cspan%20class=magenta%20dir=rtl%3E30\u200f/08\u200f/2023%3C/span%3E.)

You wouldn't want Unicode isolates in this case.

Author: I agree that in some cases bidi isolation of formatted-parts output may be achieved e.g. via `<span dir=...>` tags, but I don't think that's universally true for every use case for formatted parts.

This is one reason why e.g. in the JS Intl.MF proposal my proposal for bidi handling includes the `"none"` strategy as an alternative to the default (and required) `"compatibility"` strategy.

Effectively, I think we should include the MessageBiDiIsolationPart by default, but allow for an implementation to provide a way to get rid of them, or just have a consumer of the output ignore all the `type: "bidiIsolation"` parts.

Reviewer: I agree that span is not the only way that bidi will be implemented. `none` is easy: one can just ignore the bidi/direction. Obviously sometimes folks will want bidi controls. Or they might want to make an attributed string, I suppose, in some UI frameworks. 

I'm reluctant to create a "bidi isolation part" because it separates the direction information from the value it is associated with. It's harder if one is doing decoration, such as building some control using various tags, not all of which are span, some of which have classes or other attributes. The only place where the isolation part is easier is stringifying the message (since one just consumes it. 

```javascript
function stringifyParts(parts) {
  let res = ''
  for (let part of parts) {
    res += bidiStrategyImpl.start(part.dir);  // none does nothing, compatibility does a control, etc.
    if (part.type === 'fallback') res += `{${part.source}}`
    else if ('value' in part) res += String(part.value)
    else if ('parts' in part) {
      for (let sub of part.parts) res += String(sub.value)
    }
    res += bidiStrategyImpl.end(part.dir);
  }
  return res
}
```

Author: Doing as you propose would require implementations to provide explicit `bidiStrategyImpl.start()` and `bidiStrategyImpl.end()` interfaces, or the behaviour of format-to-parts-to-string could not be guaranteed to match the behaviour of format-to-string.

That seems like a much bigger ask than defining what the bidi isolation parts would look like if they are to be included in the parts output by the requested bidi isolation strategy.

Reviewer: Format-to-string's output would depend on what the bidi isolation strategy was, no? 

The start/end thing was just an example. Note that with isolation parts you have to "read ahead" to find out what is being isolated if what one is doing with the parts is decorating a control or generating HTML.

Author: > Format-to-string's output would depend on what the bidi isolation strategy was, no?

Yes.

> The start/end thing was just an example. Note that with isolation parts you have to "read ahead" to find out what is being isolated if what one is doing with the parts is decorating a control or generating HTML.

Sure, but either the bidi isolation parts need to be included, or some way of obtaining their equivalent is required.

Some of the MessageExpressionPart instances may be further defined
without reference to the function registry.

Unannotated expressions with a _literal_ operand
are represented by MessageStringPart.
As with MessageTextPart,

> the `value` of MessageStringPart is always a string.

Reviewer: Is there a need for the distinction? It is simply the source?

Author: In addition to `source` the `dir` and `locale` are not present on MessageTextPart.

Reviewer: Why don't text parts have a direction and language? Are they inheriting from the message?

Aside: we don't necessarily want to span every section of text, that is, this is better:

```html
<span lang=foo dir=ltr>blah blah blah 
   <span lang=bar dir=auto>placeholder here</span> blah blah</span>
```

than this:

```html
<span lang=foo dir=ltr>blah blah blah </span>
   <span lang=bar dir=auto>placeholder here</span><span lang=foo dir=ltr> blah blah</span>
```

... which is why text parts would inherit language and base paragraph direction from the message.

Author: Text parts do inherit the language and direction from the message, so they'll be whatever the formatter was called with. But they're not necessary to be repeated on every part, because the caller already knows what they asked for, and they may well be formatting the message in a context with e.g. `<body lang=foo dir=ltr>` setting them at a much higher level.

```ts
interface MessageStringPart {
  type: "string";
  source: string;
  value: string;
  dir?: "ltr" | "rtl" | "auto";
  locale?: string;
}
```

Unannotated expressions with a _variable_ operand

> that is not formatted according to its type

Reviewer: Do you mean:

```suggestion
whose type is not recognized by the implementation
or for which no default formatter is available
```

are represented by MessageUnknownPart.

```ts
interface MessageUnknownPart {
  type: "unknown";
  source: string;
  value: unknown;
}
```

When the resolution or formatting of a placeholder fails,
it is represented in the output by MessageFallbackPart.

> No `value` is provided; when formatting to a string,

Reviewer: Should we provide the fallback value? I think we have some text in the spec that allows implementations or functions to supply their own fallback.

---

Question: should a goal be that the string output of a message be equivalent to concatenating the string representation of its parts? Or at least that a test be that one can assemble the string output from the parts?

Author: > Should we provide the fallback value? I think we have some text in the spec that allows implementations or functions to supply their own fallback.

Fallback customization is only available for syntax and data model errors, which default to `�`. That would be used as the `source` value here.

> Question: should a goal be that the string output of a message be equivalent to concatenating the string representation of its parts? Or at least that a test be that one can assemble the string output from the parts?

I think the latter. With the current proposal, it's doable like this:

```js
function stringifyParts(parts) {
  let res = ''
  for (let part of parts) {
    if (part.type === 'fallback') res += `{${part.source}}`
    else if ('value' in part) res += String(part.value)
    else if ('parts' in part) {
      for (let sub of part.parts) res += String(sub.value)
    }
  }
  return res
}
```

Reviewer 3: So the parts are technically not formatted yet?
I can't get from the proposal what a `sub.value` is.
Can it be a date?

the part's representation would be `'{' + source + '}'`.

```ts
interface MessageFallbackPart {
  type: "fallback";
  source: string;
}
```

Formatting functions defined in the registry

> will need accompanying formatted-parts representations.

Reviewer: This sentence confused me, since, when I read it, I thought it was requiring a "parts" interface or that you were using too strong a word with "will". I think what you really mean is:

```suggestion
Each function defined in the registry MUST define its "formatted-parts" representation.
A function can define either a unitary string `value` or a `parts` representation.
Where possible, a function SHOULD provide a `parts` representation if its output 
might reasonably consist of multiple fields.
```

(I added the sentences after the first as a further suggestion)

Where available, such a formatted value should itself be represented by `parts`
rather than a unitary string `value`.
These sub-parts should not need fields beyond their `type` and `value`,
and in most cases it's presumed that the sub-part `value` would be a string.

```ts

> interface MessageDateTimePart {

Reviewer: Can we define the "parts" so that they are generic rather than each type having its own special part type?

Author: Yes, that's the MessageExpressionPart definition above, which these interfaces also match. The definitions here are giving more specificity about what e.g. `:datetime` and `:number` end up producing, i.e. that they have explicit `type` identifiers and define `parts` rather than `value`.

  type: "datetime";
  source: string;

>   parts: Iterable<{ type: string; value: unknown }>;

Reviewer: Notice in my example earlier in this review that I exposed the field name as a field in the parts. This becomes important when trying to decorate an Iterable whose contents shift around due to the locale/localized formatting. Dates have this feature (YMD, DMY, MDY). So do currency values (which may or may not have a decimal part, may have the symbol first or last, and may or may not have a space around the symbol). That's how the screen shots of currency values (from amazon.com and amazon.fr) get decorated.

Author: Yes, that's the `"type"` field here. I picked that rather than `"name"` because it's used by the JS [Intl formatters](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat/formatToParts#description).

Reviewer: That was super non-obvious to me (hence this conversation!), particularly since the `type` fields in the examples seemed to be focused on the "type" of formatter (datetime, number) rather than on the parts field within them. Admittedly, the MF2-level parts will be at the placeholder level. Interior parts are the problem of the formatter. But this was not at all clear and probably could use an example.

Author: Let's make sure that it's explained in more detail in the spec PR, should this design doc be accepted.

# Formatted Parts

Status: **Proposed**

<details>
	<summary>Metadata</summary>
	<dl>
		<dt>Contributors</dt>
		<dd>@eemeli</dd>
		<dt>First proposed</dt>
		<dd>2023-08-29</dd>
		<dt>Pull Request</dt>
		<dd>#000</dd>
	</dl>
</details>

## Objective

Messages often include placeholders that,

> when formatted, contain internal structure ("parts")
> that the caller might want access to
> for the purposes of styling, presentation, or manipulation.

Reviewer 4: ```suggestion
when formatted, contain internal structure ("parts").
Preserving this structure in a formatted message
may be helpful to the caller, who can then
manipulate the parts.
For example, a caller may want to style or present
messages with the same content differently
if those messages have different internal structure.
```

This proposal defines a formatted-parts target for MessageFormat 2.

## Background

Past examples have shown us that if we don't provide a formatter to parts,

> the string output will be re-parsed and re-processed by users.

Reviewer 4: It might be useful to explain why it's undesirable for users to re-parse a string (or cite some other document that explains it).

Author: ```suggestion
the string output will be re-parsed and re-processed by users.
Recent examples of web browsers needing to account for such user behaviour are available from
[June 2022](https://github.com/WebKit/WebKit/commit/1dc01f753d89a85ee19df8e8bd75f4aece80c594) and
[November 2022](https://bugs.chromium.org/p/v8/issues/detail?id=13494).
```

The shape of the formatted-parts output is defined in a manner similar to the data model,
which includes TypeScript, JSON Schema, and XML DTD definitions of the same data structure.

At the top level, the formatted-parts result is an iterable sequence of parts.
Parts corresponding to each _text_ can be simpler than those of _expressions_,
as they do not have a `source` other than their `value`,
or set any of the other possible metadata fields.

```ts
type MessageParts = Iterable<
  MessageTextPart | MessageExpressionPart | MessageBiDiIsolationPart
>;

interface MessageTextPart {
  type: "text";
  value: string;
}
```

For MessageExpressionPart, the `source` corresponds to the expression's fallback value.
The `dir` and `locale` attributes of a part may be inherited from the message
or from the operand (if present),
or overridden by an expression attribute or formatting function,
or otherwise set by the implementation.
Each part should have at most one of `value` or `parts` defined;
some may have none.

```ts
interface MessageExpressionPart {
  type: string;
  source: string;
  parts?: Iterable<{ type: string; value: unknown }>;

>   value?: unknown;

Reviewer 4: Can this be something like:

```
contents?: <Iterable<{ type: string; value: unknown }> | unknown>;
```
to capture that `parts` and `value` can't both be present?

Author: I don't think so. Using a single field for both `parts` and `value` would mean that it would be possible for a unitary value to masquerade as a sequence of parts, and be treated as such.

The shape of the formatted-parts output is defined in a manner similar to the data model,
which includes TypeScript, JSON Schema, and XML DTD definitions of the same data structure.

At the top level, the formatted-parts result is an iterable sequence of parts.
Parts corresponding to each _text_ can be simpler than those of _expressions_,
as they do not have a `source` other than their `value`,
or set any of the other possible metadata fields.

```ts
type MessageParts = Iterable<
  MessageTextPart | MessageExpressionPart | MessageBiDiIsolationPart
>;

interface MessageTextPart {
  type: "text";
  value: string;
}
```

For MessageExpressionPart, the `source` corresponds to the expression's fallback value.
The `dir` and `locale` attributes of a part may be inherited from the message
or from the operand (if present),
or overridden by an expression attribute or formatting function,
or otherwise set by the implementation.
Each part should have at most one of `value` or `parts` defined;
some may have none.

```ts
interface MessageExpressionPart {
  type: string;
  source: string;
  parts?: Iterable<{ type: string; value: unknown }>;
  value?: unknown;

>   dir?: "ltr" | "rtl" | "auto";

Reviewer 3: It is unclear to me why have `dir` and `locale` at this level.

We don't need a locale to format anything, because the parts should be already formatted.
The whole proposal is called "Formatted Parts"

The locale might be needed to render things.

Or to process the formatting result (fix grammatical agreements as a post-step, fix "a apple" to "an apple" (en) or "La abeille" => "L'abbeile" (fr), or to sentence case the result of "{item} is foo"

But then that is something that is needed for the whole collection of parts, not on `MessageExpressionPart` only.

Author: These are needed to allow embedding content in a message that uses a different script or locale than the surrounding message.

Reviewer 3: Kind of clear what they can be used for.

But this info is on the `MessageExpressionPart`, which comes from an expression.
And an expression can't create this info out of nothing, it is probably something we passed as a parameter.
So if I already know the info (because I passed it to the expression), having it on the `MessageExpressionPart` is useless duplication.

Author: Consider this message, intended for consumption by a text-to-speech system:
```
In French, the number {98 :number} is commonly expressed as {98 :number @locale=fr},
but in Belgium it's {98 :number @locale=fr-BE}.
```

How would you propose that the locale information is transmitted, if not as a field on the formatted parts?

```js
[
  { type: 'text', value: 'In French, the number ' },
  { type: 'number', source: '|98|', parts: [{ type: 'integer', value: '98' }] },
  { type: 'text', value: ' is commonly expressed as ' },
  { type: 'number', source: '|98|', locale: 'fr', parts: [{ type: 'integer', value: '98' }] },
  { type: 'text', value: ", but in Belgium it's " },
  { type: 'number', source: '|98|', locale: 'fr-BE', parts: [{ type: 'integer', value: '98' }] },
  { type: 'text', value: '.' }
]
```

As context, the `fr` number would be "quatre-vingt-dix-huit", while in `fr-BE` it's "nonante-huit".

Reviewer: Language and direction are needed for placeholders because they represent values being inserted into the overall string. The language (locale) is used to ensure proper rendering and processing (such as line breaking, text transforms, or spell-checking). The direction is used to enable bidi isolation and get the direction of the substring correct.

The language and direction of a formatted part might not match that of the message due to resource fallback when looking up the message. Or because values passed in have different language or direction. (And we want bidi isolation even if the directions match!!!!)

Providing the fields in the formatted part structure allows the user to easily access the values, e.g. it makes it easy to do something like this, resulting in proper isolation of the formatted parts (not shown is decorating the parts separately):

```javascript
var message = // whatever the host node is for the string
for (let part of formattedMessage.parts) {
    var span = document.createElement('span');
    span.lang = part.lang;
    span.dir = part.dir;
    span.appendChild(document.createTextNode(part.value));
    message.appendChild(span);
}
```

That is done on portions of the message. The whole message also has a language (locale) and base paragraph direction.
